autoscheduler/frontend/src/redux/reducers.ts

```diff
@@ -61,10 +65,120 @@ function courseCards(
+): Availability[] {
+  switch (action.type) {
+    case ADD_AVAILABILITY:
+    case MERGE_AVAILABILITY: {
+      // helper functions to get the start and end times of an availability object in minutes
+      const getStart = (av: Availability): number => av.startTimeHours * 60 + av.startTimeMinutes;
+      const getEnd = (av: Availability): number => av.endTimeHours * 60 + av.endTimeMinutes;
+
+      // check for overlaps between the availability to be added and pre-existing ones
+      let overlapsFound = 0; // counts merging overlaps only
+      let overlapIdx = -1;
+
+      const initNewAv = action.type === ADD_AVAILABILITY
+        ? argsToAvailability(action.availability)
+        : state[state.length - 1];
+      let newAv = initNewAv;
+      return state.reduce<Availability[]>((avsList, oldAv, idx): Availability[] => {
+        // only counts as an overlap if they're on the same day of the week
+        if (oldAv.dayOfWeek === newAv.dayOfWeek) {
+          const avWithEarlierStart = getStart(oldAv) < getStart(newAv) ? oldAv : newAv;
+          const avWithLaterStart = avWithEarlierStart === oldAv ? newAv : oldAv;
+          // overlap detected
+          if (getEnd(avWithEarlierStart) >= getStart(avWithLaterStart)) {
+            // if they're the same type, merge them
+            if (oldAv.available === newAv.available) {
+              overlapsFound += 1;
+              const oldOverlapIdx = overlapIdx;
+              overlapIdx = idx;
+              const newEnd = Math.max(getEnd(oldAv), getEnd(newAv));
+              const newEndHrs = Math.floor(newEnd / 60);
+              const newEndMins = newEnd % 60;
+              newAv = {
+                available: oldAv.available,
+                dayOfWeek: oldAv.dayOfWeek,
+                startTimeHours: avWithEarlierStart.startTimeHours,
+                startTimeMinutes: avWithEarlierStart.startTimeMinutes,
+                endTimeHours: newEndHrs,
+                endTimeMinutes: newEndMins,
+              };
+              if (overlapsFound === 2) {
+                return avsList.slice(0, oldOverlapIdx)
+                  .concat(avsList.slice(oldOverlapIdx + 1))
+                  .concat(newAv);
+              }
+              return avsList.filter((av) => av !== oldAv).concat(newAv);
+            }
+            // if they're different types, then set the new one to the old one's borders
+            // TODO
+          }
+        }
+        // if no overlap
+        return avsList.concat(oldAv);
+      }, []).concat(overlapsFound ? [] : initNewAv);
+    }
+    case DELETE_AVAILABILITY:
+      // filters the availability list for the availabiltiy matching the action args
+      return state.filter((av) => av.available !== action.availability.available
+        || av.dayOfWeek !== action.availability.dayOfWeek
+        || av.startTimeHours * 60 + av.startTimeMinutes !== action.availability.time1
+        || av.endTimeHours * 60 + av.endTimeMinutes !== action.availability.time2);
```

Comment:
Yes, it took me a second, but there actually is a reason. In the case of `DELETE_AVAILABILITY`, it is known that the argument `time1` refers to the start time and `time2` refers to the end time. However, for `UPDATE_AVAILABILITY`, `time1` could match either the start time or the end time. 

Therefore, since this giant conditional will return `false` iff `avl` does not match the arguments, in the case of deleting, finding a single un-matching time means the whole availability does not match, but in the case of updating, it must be verified that neither start nor end time matches in order to safely move onto the next availability.